feat(ProjectGroupFilter):Filter the projects in Advanced Search based on Projects Group.

ravi110336 opened the pull request:
[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> The Filter for Group in Advanced Search should read from the Project List Group not from the user table.
> * Which issue is this pull request belonging to and how is it solving it? (#1194 )
> * Did you add or update any new dependencies that are required for your change? No

Issue: closes #1194 

### Suggest Reviewer
> You can suggest reviewers here with an @mention.

### How To Test?

- To Test this, go to the Projects tab Advanced Search option and select the filter by Group and search the projects.
- The values populated in the Group in Advanced Search is the distinct Group from the list of projects available in the DB not 
      from the user table.

> Have you implemented any additional tests? No

### Checklist
Must:
- [x] All related issues are referenced in commit messages and in PR

Signed-off-by: ravi110336 <kumar.ravindra@siemens.com>

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortalConstants.java:
```diff
@@ -581,6 +581,8 @@
 
     public static final String PREDEFINED_TAGS;
 
+    public static final String PROJECT_GROUPS = "projectGroups";
```
Maybe the same variable can be reused, instead of creating a new one
`PortalConstants.ORGANIZATIONS`

ravi110336 replied:
Used the same variable.

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -1212,9 +1212,11 @@ public void doView(RenderRequest request, RenderResponse response) throws IOExce
-        List<Organization> organizations = UserUtils.getOrganizations(request);
-        request.setAttribute(PortalConstants.ORGANIZATIONS, organizations);
+        List<Project> projects = getFilteredProjectList(request);
+        projectGroups = getProjectGroups(projects);
```
No need to for line 1215 and can be written as 
`Set<String> projectGroups = getProjectGroups(projects);`
because `getProjectGroups()` handles empty set of groups

ravi110336 replied:
Fixed, Please check.

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -2692,4 +2694,18 @@ private void createModRequest(ResourceRequest request, ResourceResponse response
     }
+
+    private Set<String> getProjectGroups(List<Project> projects) {
+        Set<String> businessUnit = new HashSet<String>();
```
Better to have `TreeSet` instead of `HashSet` since it offers alphabetic sorting

ravi110336 replied:
Used TreeSet for sorting.

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -2692,4 +2694,18 @@ private void createModRequest(ResourceRequest request, ResourceResponse response
+                    businessUnit.add(project.getBusinessUnit());
+                }
+            }
+        } catch (Exception ex) {
```
Is it required to catch Exception because there is no checked Exception thrown.
Also this process is Primary process, and not a secondary process like `logging`, `changelogs` etc.

ravi110336 replied:
Removed the try catch block.

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -2692,4 +2694,18 @@ private void createModRequest(ResourceRequest request, ResourceResponse response
+    private Set<String> getProjectGroups(List<Project> projects) {
+        Set<String> businessUnit = new HashSet<String>();
+        try {
+            if(!CommonUtils.isNullOrEmptyCollection(projects)) {
```
Maybe this can be used `CommonUtils.isNotEmpty(projects)`

ravi110336 replied:
Fixed please check.

JaideepPalit reviewed (changes requested):
**Reviewed the PR.**
Kindly have a look at some review comments

**Tested the PR**
Seems to be some issue.
Steps to reproduce.
1.Before search the list is coming properly, 
![before_search](https://user-images.githubusercontent.com/56534710/116852499-09477980-ac12-11eb-9316-670588fd3492.JPG)

2. Select any one Group and search.
3. After search only selected groups is present in the list , other are not there
![after_search](https://user-images.githubusercontent.com/56534710/116852512-0f3d5a80-ac12-11eb-9ef1-bf7b4714c691.JPG)
4. Same issue persists when searched with multiple filters in combination.
![combination](https://user-images.githubusercontent.com/56534710/116852567-2c722900-ac12-11eb-8372-7a5d737f7ec9.JPG)

ravi110336 commented:
Thanks @JaideepPalit for the review comments and testing, incorporated all the changes and fixed the issue with the search list, Please check and let me know in-case of any issue.

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/common/PortalConstants.java:
```diff
@@ -581,6 +581,7 @@
 
     public static final String PREDEFINED_TAGS;
 
+
```
Kindly remove unused line

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -2692,4 +2699,14 @@ private void createModRequest(ResourceRequest request, ResourceResponse response
+
+    private Set<String> getProjectGroups(List<Project> projects) {
+        Set<String> businessUnit = new TreeSet<String>();
+            if(CommonUtils.isNotEmpty(projects)) {
```
Indentation

JaideepPalit commented on frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/portlets/projects/ProjectPortlet.java:
```diff
@@ -1212,13 +1212,20 @@ public void doView(RenderRequest request, RenderResponse response) throws IOExce
+                String parameter = request.getParameter(filteredField.toString());
+                request.setAttribute(filteredField.getFieldName(), nullToEmpty(parameter));
+            }
+        }catch(TException e) {
```
Indentation. `}catch`
Also kindly don't put entire method in try-catch.
I guess only this part should be placed in try-catch - `projectList = projectClient.getAccessibleProjectsSummary(user);`

JaideepPalit reviewed (changes requested):
Tested the PR. Issue is Fixed and working fine.

ravi110336 commented:
Thanks @JaideepPalit for the review comments, incorporated the changes please check.

JaideepPalit reviewed (approved):
Code looks fine to me.
